### PR TITLE
Query param decoding

### PR DESCRIPTION
**Pull request by sibisimon**
#### Neo4j Import
neo4j

#### Dev Tools
dev-tools

[INVE-274]: https://sirensolutions.atlassian.net/browse/INVE-274

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -155,12 +155,23 @@ internals.handler = function (route, handlerOptions) {
...
             // Encoding request path
-            const _encodeURL = (url) => {
-                if (decodeURI(url) === url) {
-                    url = encodeURI(url);
+            const _encodeURL = (uri) => {
+                let uriBasePath = uri;
+                let haveQueryParam = false;
+                let queryParams = '';
+
+                if (uri.match(/\?./)) {
```
Comment (szydan) [inline lib/index.js]: Have a look at this commit 
We will likely merge it soon 
I used URL nodejs package to split the url into parts 
https://github.com/sirensolutions/h2o2/commit/82714624ffe3ae6623b01a3c5d00443a21614288

```
const URL = require('url');
```
It is best to use this native package then to write a custom code for parsing urls 
Could you also point me to the issue where the original bug was described ?

@johnnylopes  Maybe we should think about integration test for ingestion which would cover the scenario from the original ticket (ask @sibisimon) 
and another one to cover the investigate dev tools - as the fix done initially broke them (but only in dev mode so I do not know how we could test that)

**Review by szydan — CHANGES_REQUESTED** [review]

File: lib/index.js (hunk 1 above)
Comment (sibisimon) [reply lib/index.js]: https://sirensolutions.atlassian.net/browse/INVE-274

File: lib/index.js [hunk 2]
```diff
@@ -155,21 +156,16 @@ internals.handler = function (route, handlerOptions) {
...
-                if (uri.match(/\?./)) {
-                    haveQueryParam = true;
-                    uriBasePath = uri.split('?')[0];
-                    queryParams = uri.split('?')[1];
+            const _encodeURL = (url) => {
+                let encodedUrl = url;
+                let basePath = '';
+                const { pathname, protocol, host, search } = URL.parse(url);
+
+                if (decodeURI(pathname) === pathname) {
```
Comment (szydan) [inline lib/index.js]: Why not 

```javascript
...
const uri = URL.parse(url)
if (uri.pathname && decodeURI(uri.pathname) === uri.pathname) {
  uri.pathname = encodeURI(uri.pathname)
  return URL.format(uri)
}
return url;
```

**Review by szydan — CHANGES_REQUESTED** [review]

**Review by szydan — APPROVED** [review]
The change is fine for me 
but Can we here a comment why exactly we have to do it 
Just for the record to understand the reason

**Comment by sibisimon** [comment]
### Issue

If we have unescaped characters (eg. Arabic characters ) in our pathname, our server APIs will fail with an error. So, the reason for it is, our proxy will decode pathname but it didn't encode it before sending it to our server (Only path, the query params seems fine).

#### Example: 
Url having special characters send from the UI: https://localhost:5606/ith/ingest/api/v1/clearJobLogs/neo4j--url-test:%D8%A7%D9%84%D9%85%D9%88%D8%B1%D8%AF 

But the path is decoded in our proxy handler and the request sent to our server is: https://localhost:5603/ingest/api/v1/clearJobLogs/neo4j--url-test:الفئة

So, we are encoding this path again before sending it.

**Comment by szydan** [comment]
@fabiocorneti here is a fix by @sibisimon 
You wanted to have a look there is an example and description in the comment above

**Pull request merged**